flink-streaming-java/src/test/java/org/apache/flink/streaming/runtime/tasks/SourceStreamTaskTest.java

```diff
@@ -479,5 +481,40 @@ public void cancel() {
 			running = false;
 		}
 	}
+
+	private static final class OutputRecordInCloseTestSource extends RichSourceFunction<String> {
+
+		private final String name;
+		private final String[] elements;
+
+		private transient volatile boolean canceled = false;
+
+		private transient SourceContext<String> context;
```

Comment:
we can remove `canceled` and `context` for simple.

Reply:
We need to use `context` to output data in `close`, but `canceled` can be removed.